### PR TITLE
FOUNDATIONS: Add Language

### PR DESCRIPTION
**Pull request by ElbekDeveloper**

### CODE REVIEW

File: Sofee.Core.Api.Tests.Unit/Services/Foundations/LanguageServiceTests.Validation.Add.cs [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+            //then
+            await Assert.ThrowsAsync<LanguageValidationException>(() =>
+                addLanguageTask.AsTask());
+
+            this.storageBrokerMock.Verify(broker =>
+                   broker.InsertLanguageAsync(nullLanguage),
+                       Times.Never);
+
+            this.loggingBrokerMock.Verify(broker =>
+                broker.LogCritical(It.Is(SameExceptionAs(
```
Comment (ElbekDeveloper) [inline Sofee.Core.Api.Tests.Unit/Services/Foundations/LanguageServiceTests.Validation.Add.cs]: @hassanhabib this wrong logging method call was corrected in pass commit. Updated LogCritical to LogError

File: Sofee.Core.Api.Tests.Unit/Services/Foundations/LanguageServiceTests.Logic.Add.cs [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+namespace Sofee.Core.Api.Tests.Unit.Services.Foundations
+{
+    public partial class LanguageServiceTests
+    {
+        [Fact]
+        public async Task ShouldAddLanguageAsync()
+        {
+            //given 
+            DateTimeOffset dateTime = GetRandomDateTime();
```
Comment (ShriHumrudha) [inline Sofee.Core.Api.Tests.Unit/Services/Foundations/LanguageServiceTests.Logic.Add.cs:22]: Do we need this variable?

File: Sofee.Core.Api.Tests.Unit/Services/Foundations/LanguageServiceTests.Validation.Add.cs [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+            //then
+            await Assert.ThrowsAsync<LanguageValidationException>(() =>
+                addLanguageTask.AsTask());
+
+            this.storageBrokerMock.Verify(broker =>
+                   broker.InsertLanguageAsync(nullLanguage),
+                       Times.Never);
+
+            this.loggingBrokerMock.Verify(broker =>
```
Comment (ShriHumrudha) [inline Sofee.Core.Api.Tests.Unit/Services/Foundations/LanguageServiceTests.Validation.Add.cs:40]: The logging broker will be verified before the storage broker. The same goes with the 'VerifyNoOtherCalls' order.

File: Sofee.Core.Api.Tests.Unit/Services/Foundations/LanguageServiceTests.cs [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+            this.languageService = new LanguageService(
+                storageBroker: this.storageBrokerMock.Object,
+                loggingBroker: this.loggingBrokerMock.Object);
+        }
+
+        private static Language CreateRandomLanguage() =>
+                CreateLanguageFiller().Create();
+
+        private static Filler<Language> CreateLanguageFiller()
```
Comment (ShriHumrudha) [inline Sofee.Core.Api.Tests.Unit/Services/Foundations/LanguageServiceTests.cs:37]: This method to be the last in this class.

File: Sofee.Core.Api.Tests.Unit/Services/Foundations/LanguageServiceTests.cs [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+        {
+            var filler = new Filler<Language>();
+            Guid createdById = Guid.NewGuid();
+
+            filler.Setup()
+                .OnProperty(language => language.CreatedDate).Use(GetRandomDateTime())
+                .OnProperty(language => language.UpdatedDate).IgnoreIt()
+                .OnProperty(language => language.CreatedBy).Use(createdById)
+                .OnProperty(language => language.UpdatedBy).IgnoreIt();
```
Comment (ShriHumrudha) [inline Sofee.Core.Api.Tests.Unit/Services/Foundations/LanguageServiceTests.cs:46]: Curious to know, why we ignore updated fields.

File: Sofee.Core.Api/Models/Languages/Exceptions/LanguageValidationException.cs [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+
+using Xeptions;
+
+namespace Sofee.Core.Api.Models.Languages.Exceptions
+{
+    public class LanguageValidationException : Xeption
+    {
+        public LanguageValidationException(Xeption innerException)
+            : base(message: "Language validation error occured. Please, try again.", 
+                       innerException: innerException)
```
Comment (ShriHumrudha) [inline Sofee.Core.Api/Models/Languages/Exceptions/LanguageValidationException.cs:14]: The parameter aliasing can be skipped here.

File: Sofee.Core.Api/Services/Foundations/Languages/ILanguageService.cs [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+// FREE TO USE FOR THE WORLD
+// -------------------------------------------------------
+
+using Sofee.Core.Api.Models.Languages;
+using System.Threading.Tasks;
+
+namespace Sofee.Core.Api.Services.Foundations.Languages
+{
+    public interface ILanguageService
+{
```
Comment (ShriHumrudha) [inline Sofee.Core.Api/Services/Foundations/Languages/ILanguageService.cs:12]: Need one more tab here.

File: Sofee.Core.Api/Services/Foundations/Languages/LanguageService.Exceptions.cs [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+            {
+                throw CreateAndLogValidationException(nullLanguageException);
+            }
+        }
+
+        private Xeption CreateAndLogValidationException(
+            NullLanguageException nullLanguageException)
+        {
+            var languageValidationException =
+             new LanguageValidationException(nullLanguageException);
```
Comment (ShriHumrudha) [inline Sofee.Core.Api/Services/Foundations/Languages/LanguageService.Exceptions.cs:33]: Fix the indendation.

**Comment by ShriHumrudha** [comment]
@ElbekDeveloper We may need a few more tests to take this PR to closure.

**Review by ShriHumrudha — CHANGES_REQUESTED** [review]

File: Sofee.Core.Api.Tests.Unit/Services/Foundations/LanguageServiceTests.Validation.Add.cs (hunk 3 above)
Comment (ElbekDeveloper) [reply Sofee.Core.Api.Tests.Unit/Services/Foundations/LanguageServiceTests.Validation.Add.cs:40]: done

File: Sofee.Core.Api.Tests.Unit/Services/Foundations/LanguageServiceTests.Logic.Add.cs (hunk 2 above)
Comment (ElbekDeveloper) [reply Sofee.Core.Api.Tests.Unit/Services/Foundations/LanguageServiceTests.Logic.Add.cs:22]: deleting it

File: Sofee.Core.Api.Tests.Unit/Services/Foundations/LanguageServiceTests.cs (hunk 4 above)
Comment (ElbekDeveloper) [reply Sofee.Core.Api.Tests.Unit/Services/Foundations/LanguageServiceTests.cs:37]: done

File: Sofee.Core.Api.Tests.Unit/Services/Foundations/LanguageServiceTests.cs (hunk 5 above)
Comment (ElbekDeveloper) [reply Sofee.Core.Api.Tests.Unit/Services/Foundations/LanguageServiceTests.cs:46]: I thought on adding the profile, update fields will have no value. So, thought we can ignore them

File: Sofee.Core.Api/Models/Languages/Exceptions/LanguageValidationException.cs (hunk 6 above)
Comment (ElbekDeveloper) [reply Sofee.Core.Api/Models/Languages/Exceptions/LanguageValidationException.cs:14]: done

File: Sofee.Core.Api/Services/Foundations/Languages/ILanguageService.cs (hunk 7 above)
Comment (ElbekDeveloper) [reply Sofee.Core.Api/Services/Foundations/Languages/ILanguageService.cs:12]: done

File: Sofee.Core.Api/Services/Foundations/Languages/LanguageService.Exceptions.cs (hunk 8 above)
Comment (ElbekDeveloper) [reply Sofee.Core.Api/Services/Foundations/Languages/LanguageService.Exceptions.cs:33]: done

File: Sofee.Core.Api.Tests.Unit/Services/Foundations/LanguageServiceTests.cs (hunk 5 above)
Comment (ShriHumrudha) [reply Sofee.Core.Api.Tests.Unit/Services/Foundations/LanguageServiceTests.cs:57]: When we add the profile, even the updated date / by will take a value similar to the created date / by

Comment (ElbekDeveloper) [reply Sofee.Core.Api.Tests.Unit/Services/Foundations/LanguageServiceTests.cs:57]: Now, I got it

**Pull request merged**